envoy/config/config_validator.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include "envoy/common/optref.h"
+#include "envoy/config/subscription.h"
+#include "envoy/server/instance.h"
+
+#include "source/common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * The configuration validator interface. One can implement such validator
+ * to add custom constraints to the fetched config, and reject a config
+ * which violates these constraints.
+ * The validators will be extensions that can be dynamically configured.
+ *
+ * A validator example: a validator that prevents removing all Clusters
+ * from an Envoy (which may be caused by a bug in the config plane, and not
+ * intentional change).
+ */
+class ConfigValidator {
+public:
+  virtual ~ConfigValidator() = default;
+
+  /**
+   * Validates a given set of resources matching a State-of-the-World update.
+   * @param server A server instance to fetch the state before applying the config.
+   * @param resources List of decoded resources that reflect the new state.
+   * @throw optional EnvoyException if the config should be rejected.
+   * @return whether the config is valid or not.
+   */
+  virtual bool validate(Server::Instance& server,
+                        const std::vector<DecodedResourceRef>& resources) PURE;
+
+  /**
+   * Validates a given set of resources matching an Incremental update.
+   * @param server A server instance to fetch the state before applying the config.
+   * @param added_resources A list of decoded resources to add to the current state.
+   * @param removed_resources A list of resources to remove from the current state.
+   * @throw optional EnvoyException if the config should be rejected.
+   * @return whether the config is valid or not.
+   */
+  virtual bool validate(Server::Instance& server,
```

Comment:
The server instance is pretty wide, is there a more useful abstraction to be able to understand a particular xDS config that we can offer? The answer might be "no" and maybe we should offer some libraries to help folks take a `Server` and get to this state. The anti-pattern I worry about is that validators might get quite complicated just in the part that needs to understand current state.

Reply:
The first version, as was specified in the design doc, defined a new struct that included the ClusterManager and ListenerManager. Passing `Server::Instance` is the most generic thing to pass, and I agree that it makes "everything" accessible, which we might want to avoid.
I guess we can start with the struct that has only the cluster and listener managers, and add other things if needed in the future (e.g., for validation of SDS, or RTDS).